subsys/sip_svc/sip_svc.c

```diff
@@ -0,0 +1,898 @@
+	LOG_DBG("\ta1             %08lx\n", a1);
+	LOG_DBG("\ta2             %08lx\n", a2);
+	LOG_DBG("\ta3             %08lx\n", a3);
+	LOG_DBG("\ta4             %08lx\n", a4);
+	LOG_DBG("\ta5             %08lx\n", a5);
+	LOG_DBG("\ta6             %08lx\n", a6);
+	LOG_DBG("\ta7             %08lx\n", a7);
+
+	ctrl->invoke_fn(a0, a1, a2, a3, a4, a5, a6, a7, &res);
+
+	LOG_DBG("after  %s (polling async response)\n", ctrl->method);
+	LOG_DBG("\ta0             %08lx\n", res.a0);
+	LOG_DBG("\ta1             %08lx\n", res.a1);
+	LOG_DBG("\ta2             %08lx\n", res.a2);
+	LOG_DBG("\ta3             %08lx\n", res.a3);
+
+	/* Callback if get response */
+	ret = sip_svc_plat_async_res_res(&res,
+			ctrl->async_resp_data, &data_size,
+			&trans_id);
+
+	if (ret == -ENOTSUP) {
+		return 0;
+	}
+
+	if (ret != 0) {
+		return -EINPROGRESS;
+	}
+
+	/* get caller information based on trans id */
+	trans_id_item = sip_svc_id_map_query_item(ctrl->trans_id_map,
+						     trans_id);
+
+	if (!trans_id_item) {
+		return -ENOENT;
+	}
+
+	header = SIP_SVC_PROTO_HEADER(sip_svc_plat_get_error_code(&res),
+				      trans_id);
+
+	/* Get caller provided memory space to put response */
+	data_addr = (((uint64_t) trans_id_item->arg2) |
+		     ((uint64_t) trans_id_item->arg3) << 32);
+
+	/* Check caller provided memory space to avoid overflow */
+	if (data_size > ((size_t) trans_id_item->arg4))
+		data_size = ((size_t)trans_id_item->arg4);
+
+	response.header = header;
+	response.a0 = res.a0;
+	response.a1 = res.a1;
+	response.a2 = res.a2;
+	response.a3 = res.a3;
+	response.resp_data_addr = data_addr;
+	response.resp_data_size = data_size;
+	response.priv_data = trans_id_item->arg5;
+
+	/* Copy async cmd response into caller given memory space */
+	if (data_addr)
+		memcpy((char *)data_addr, ctrl->async_resp_data, data_size);
+
+	sip_svc_callback(ctrl,
+			    trans_id,
+			    (char *)&response,
+			    sizeof(struct sip_svc_response));
+
+	/* Check again is there any async busy job id */
+	ctrl->active_job_cnt--;
+	ctrl->active_async_job_cnt--;
+	if (ctrl->active_async_job_cnt == 0)
+		return 0;
+
+	return -EINPROGRESS;
+}
+
+static void sip_svc_thread(void *ctrl_ptr, void *arg2, void *arg3)
+{
+	ARG_UNUSED(arg2);
+	ARG_UNUSED(arg3);
+
+	struct sip_svc_controller *ctrl = (struct sip_svc_controller *)ctrl_ptr;
+	int ret_msgq;
+	int ret_resp;
+
+	while (1) {
+		ret_msgq = -EINPROGRESS;
+		ret_resp = -EINPROGRESS;
+		while (ret_msgq != 0 || ret_resp != 0) {
+			ret_msgq = sip_svc_request_handler(ctrl);
+			ret_resp = sip_svc_async_response_handler(ctrl);
+		}
+		k_thread_suspend(ctrl->tid);
+	}
+}
+
+int sip_svc_send(struct sip_svc_controller *ctrl,
+		    uint32_t c_token,
+		    char *sip_svc_request,
+		    int size,
+		    sip_svc_cb_fn cb)
+{
+	struct sip_svc_request *request =
+		(struct sip_svc_request *)sip_svc_request;
+	uint32_t trans_id = SIP_SVC_ID_INVALID;
+	uint32_t trans_idx = SIP_SVC_ID_INVALID;
+	uint32_t c_idx;
+
+	if (!ctrl)
+		return SIP_SVC_ID_INVALID;
```

Comment:
Again, is that better to use `return -EINVAL;` here?

Reply:
fixed to use error number return